src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp

```diff
@@ -0,0 +1,297 @@
+  const return_type rho = sqrt(square(xbar) + square(ybar));
+  const return_type sin_factor =
+      radius_ * Endcap_detail::sin_ax_over_x(rho, theta_);
+  const return_type d_sin_factor =
+      radius_ * Endcap_detail::dlogx_sin_ax_over_x(rho, theta_);
+
+  auto jacobian_matrix =
+      make_with_value<tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>>(
+          dereference_wrapper(source_coords[0]), 0.0);
+
+  // dz/dxbar
+  get<2, 0>(jacobian_matrix) = -sin_factor * theta_ * xbar;
+  // dz/dybar
+  get<2, 1>(jacobian_matrix) = -sin_factor * theta_ * ybar;
+  // dx/dxbar
+  get<0, 0>(jacobian_matrix) = d_sin_factor * square(xbar) + sin_factor;
+  // dx/dybar
+  get<0, 1>(jacobian_matrix) = d_sin_factor * xbar * ybar;
+  // dy/dxbar
+  get<1, 0>(jacobian_matrix) = d_sin_factor * ybar * xbar;
+  // dy/dybar
+  get<1, 1>(jacobian_matrix) = d_sin_factor * square(ybar) + sin_factor;
+
+  return jacobian_matrix;
+}
+
+template <typename T>
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame> Endcap::inv_jacobian(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using return_type = tt::remove_cvref_wrap_t<T>;
+  const return_type& xbar = source_coords[0];
+  const return_type& ybar = source_coords[1];
+  const return_type rho = sqrt(square(xbar) + square(ybar));
+  // Let q = sin(rho theta)/rho
+  const return_type q = Endcap_detail::sin_ax_over_x(rho, theta_);
+  const return_type dlogrho_q = Endcap_detail::dlogx_sin_ax_over_x(rho, theta_);
+  const return_type one_over_r_q = 1.0 / (q * radius_);
+  const return_type tmp =
+      one_over_r_q * dlogrho_q / (q + square(rho) * dlogrho_q);
+
+  auto inv_jacobian_matrix =
+      make_with_value<tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>>(
+          dereference_wrapper(source_coords[0]), 0.0);
+
+  // dxbar/dx
+  get<0, 0>(inv_jacobian_matrix) = one_over_r_q - square(xbar) * tmp;
+  // dxbar/dy
+  get<0, 1>(inv_jacobian_matrix) = -xbar * ybar * tmp;
+  // dybar/dx
+  get<1, 0>(inv_jacobian_matrix) = get<0, 1>(inv_jacobian_matrix);
+  // dybar/dy
+  get<1, 1>(inv_jacobian_matrix) = one_over_r_q - square(ybar) * tmp;
+
+  return inv_jacobian_matrix;
+}
+
+boost::optional<std::array<double, 3>> Endcap::inverse(
+    const std::array<double, 3>& target_coords, const double sigma_in) const
+    noexcept {
+  const double x = target_coords[0] - center_[0];
+  const double y = target_coords[1] - center_[1];
+  const double z = target_coords[2] - center_[2];
+  const double r = sqrt(square(x) + square(y) + square(z));
+  // The equal_within_roundoff below has an implicit scale of 1,
+  // so the inverse may fail if radius_ is very small on purpose,
+  // e.g. if we really want a tiny tiny domain for some reason.
+  if (not equal_within_roundoff(r, radius_)) {
+    return boost::none;
+  }
+
+  // Compute sin^2(rho theta).
+  const double sin_squared_rho_theta = (square(x) + square(y)) / square(r);
+  // Compute sin(rho theta)/rho.
```

Comment:
I think the sin^2(rho theta) calculation should be a detail function

Reply:
Don't reinvent the wheel.  Use `boost::math::sinc_pi`.

Reply:
@wthrowe Thanks!  Will use boost::math::sinc_pi.  Do you know if there's a boost function for the derivative of sync_pi (which is also problematic at x=0)?  If not, I need to keep my version.

Reply:
I do not see one for the derivative, no.

The diagonal components of your Jacobian can be evaluated naively, because the first term approaches 1 and the second term becomes small, including its numerical errors.  The off-diagonal terms I don't see any way to evaluate without a series expansion with better than an absolute error of O(eps), which can be a large relative error (although not large compared with the other components of the matrix).

I see a few other places where the math could be simplified in this class.  I'll go through it all sometime in the next few days.

Reply:
boost::math::sync_pi doesn't work out of the box for DataVectors.  Apparently to get it working one would need to add std::numeric_limits<DataVector>::bla  functions.  Or maybe there's a way with boost::math::policies.

Reply:
Also, it seems I can't pass blaze expressions into boost::math::sync_pi.

Reply:
Your Taylor expansion implementation doesn't work for DataVectors or Blaze expressions either.

But it sounds like you need the expansion of the derivative in any case, so I'm fine with implementing them both the same way.